src/box/box.cc

```diff
@@ -1805,6 +1805,214 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
+	struct fiber *waiter;
+	/** Collected vclock. */
+	struct vclock *vclock;
+	/** The request deadline. */
+	double deadline;
+	/** Whether the request is timed out. */
+	bool timed_out;
+};
+
+/**
+ * A trigger executed on each ack to collect up to date remote node vclocks.
+ * When an ack comes with requested sync for some replica id, ack vclock is
+ * accounted.
+ */
+static int
+check_sync_on_ack(struct trigger *trigger, void *event)
+{
+	struct replication_ack *ack = (struct replication_ack *)event;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = ack->source;
+	uint64_t sync = data->syncs[id];
+	if (!bit_test(&data->collected_vclock_map, id) && sync > 0 &&
+	    ack->sync >= sync) {
+		vclock_max_ignore0(data->vclock, ack->vclock);
+		bit_set(&data->collected_vclock_map, id);
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** A trigger querying relay's next sync value once it becomes operational. */
+static int
+relay_get_sync_on_start(struct trigger *trigger, void *event)
+{
+	struct replica *replica = (struct replica *)event;
+	if (replica->anon)
+		return 0;
+	struct relay *relay = replica->relay;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = replica->id;
+	try {
+		if (!bit_test(&data->collected_vclock_map, id)) {
+			data->syncs[id] = relay_next_sync(relay,
+							  data->deadline);
+		}
+	} catch (TimedOut *) {
+		data->timed_out = true;
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** Find the minimal vclock which has all the data confirmed on a quorum. */
+static int
+box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
+{
+	/*
+	 * How many vclocks we should see to be sure that at least one of them
+	 * contains all data present on any real quorum.
+	 */
+	int vclock_count = MAX(1, replicaset.registered_count) -
+			   replication_synchro_quorum + 1;
+	/*
+	 * We should check the vclock on self plus vclock_count - 1 remote
+	 * instances.
+	 */
+	vclock_copy(confirmed_vclock, &replicaset.vclock);
+	if (vclock_count <= 1)
+		return 0;
+
+	struct sync_trigger_data data = {
+		.syncs = {0},
+		.collected_vclock_map = 0,
+		.waiter = fiber(),
+		.vclock = confirmed_vclock,
+		.deadline = deadline,
+		.timed_out = false,
+	};
+	struct trigger on_relay_thread_start;
+	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
+		       NULL);
+	trigger_add(&replicaset.on_relay_thread_start, &on_relay_thread_start);
+	struct trigger on_ack;
+	trigger_create(&on_ack, check_sync_on_ack, &data, NULL);
+	trigger_add(&replicaset.on_ack, &on_ack);
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
+		    replica->anon) {
+			continue;
+		}
+		/* Might be already filled by on_relay_thread_start trigger. */
+		if (data.syncs[replica->id] == 0) {
+			data.syncs[replica->id] =
+				relay_next_sync(replica->relay, deadline);
```

Comment:
Thanks, my bad. I made it return an error instead of throwing.

Will cover it with a test tomorrow.

<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 248aab990..8a793df7d 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -1856,12 +1856,11 @@ relay_get_sync_on_start(struct trigger *trigger, void *event)
 	struct sync_trigger_data *data =
 		(struct sync_trigger_data *)trigger->data;
 	uint32_t id = replica->id;
-	try {
-		if (!bit_test(&data->collected_vclock_map, id)) {
-			data->syncs[id] = relay_next_sync(relay,
-							  data->deadline);
-		}
-	} catch (TimedOut *) {
+	/* Already accounted. */
+	if (bit_test(&data->collected_vclock_map, id))
+		return 0;
+	if (relay_next_sync(relay, &data->syncs[id], data->deadline) != 0) {
+		diag_clear(diag_get());
 		data->is_timed_out = true;
 		fiber_wakeup(data->waiter);
 	}
@@ -1907,9 +1906,14 @@ box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
 			continue;
 		}
 		/* Might be already filled by on_relay_thread_start trigger. */
-		if (data.syncs[replica->id] == 0) {
-			data.syncs[replica->id] =
-				relay_next_sync(replica->relay, deadline);
+		if (data.syncs[replica->id] != 0)
+			continue;
+		if (relay_next_sync(replica->relay, &data.syncs[replica->id],
+				    deadline) != 0) {
+			/* Timed out. */
+			trigger_clear(&on_ack);
+			trigger_clear(&on_relay_thread_start);
+			return -1;
 		}
 	}
 	replicaset_foreach(replica) {
diff --git a/src/box/relay.cc b/src/box/relay.cc
index b900100be..493c95eba 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -872,8 +872,8 @@ relay_next_sync_f(struct cbus_call_msg *msg)
 	return 0;
 }
 
-uint64_t
-relay_next_sync(struct relay *relay, double deadline)
+int
+relay_next_sync(struct relay *relay, uint64_t *sync, double deadline)
 {
 	struct relay_next_sync_msg *msg =
 		(struct relay_next_sync_msg *)xmalloc(sizeof(*msg));
@@ -882,10 +882,10 @@ relay_next_sync(struct relay *relay, double deadline)
 	if (cbus_call_timeout(&relay->relay_pipe, &relay->tx_pipe, &msg->base,
 			      relay_next_sync_f, relay_next_sync_msg_free,
 			      timeout) < 0)
-		diag_raise();
-	uint64_t sync = msg->sync;
+		return -1;
+	*sync = msg->sync;
 	free(msg);
-	return sync;
+	return 0;
 }
 
 uint64_t
diff --git a/src/box/relay.h b/src/box/relay.h
index eff2eb3d1..927a4e9c9 100644
--- a/src/box/relay.h
+++ b/src/box/relay.h
@@ -101,8 +101,8 @@ double
 relay_txn_lag(const struct relay *relay);
 
 /** Returns the next sync relay is about to send. */
-uint64_t
-relay_next_sync(struct relay *relay, double deadline);
+int
+relay_next_sync(struct relay *relay, uint64_t *sync, double deadline);
 
 /** Returns the last sync received in response from a replica. */
 uint64_t

```

</details>

Reply:
Added a test:
<details><summary>test</summary>

```diff
diff --git a/test/replication-luatest/linearizable_read_test.lua b/test/replication-luatest/linearizable_read_test.lua
index a535b76ed..d963d2b58 100644
--- a/test/replication-luatest/linearizable_read_test.lua
+++ b/test/replication-luatest/linearizable_read_test.lua
@@ -80,6 +80,18 @@ g.test_wait_others = function(g)
     t.assert(ok, 'begin() succeeds after replica connections are established')
 end
 
+g.test_timeout = function(g)
+    g.servers[1]:exec(function()
+        local t = require('luatest')
+        box.error.injection.set('ERRINJ_RELAY_FROM_TX_DELAY', true)
+        local ok, err = pcall(box.begin, {linearizable = true, timeout = 0.01})
+        box.error.injection.set('ERRINJ_RELAY_FRMO_TX_DELAY', false)
+        t.assert(not ok, 'Error when relay is unresponsive')
+        t.assert_equals(err.type, 'TimedOut',
+                        'Timeout when relay is unresponsive')
+    end)
+end
+
 g.before_test('test_no_dirty_reads', function(g)
     -- Make cluster size 5. Having quorum 3 out of 5 will make the instance
     -- poll 5 - 3 + 1 == all 3 of the "real" nodes.
```

</details>